examples/advanced/llm_hf/README.md

```diff
@@ -0,0 +1,131 @@
+Install additional requirements:
+```
+python3 -m pip install -r requirements.txt
+```
+
+## Model and Data Preparation
+We first download the model and save it to the `model` folder, note that approved access to the model is required
+```
+mkdir model
+cd model
+git clone https://huggingface.co/meta-llama/Llama-2-7b-hf
+cd ..
+```
+
+We then download and preprocess (to be consistent with our [NeMo example](../../../integration/nemo/examples/supervised_fine_tuning), we follow the same preprocessing steps) the dataset [databricks-dolly-15k](https://huggingface.co/datasets/databricks/databricks-dolly-15k) for this example. 
+```
+mkdir dataset
+cd dataset
+git clone https://huggingface.co/datasets/databricks/databricks-dolly-15k
+cd ..
+mkdir dataset/dolly
+python ./utils/preprocess_dolly.py --training_file dataset/databricks-dolly-15k/databricks-dolly-15k.jsonl --output_dir dataset/dolly
+```
+
+## Centralized Training
+### One-call training
+Centralized trainings, as the baseline for comparison with FL results, are done with the following command:
+```
+python3 ./utils/hf_sft_peft.py --output_path ./workspace_centralized/llama2-7b-dolly-sft --mode 0
+python3 ./utils/hf_sft_peft.py --output_path ./workspace_centralized/llama2-7b-dolly-peft --mode 1
+```
+### Adaptation Step 1: iterative training
+To adapt the centralized training script to federated application, we first need to "break" the single call to `trainer.train()` into iterative calls, one for each round of training.
+For this purpose, we provided `utils/hf_sft_peft_iter.py` as an example, which is a modified version of `utils/hf_sft_peft.py`.
+Their differences are highlighted below:
+
+![diff](./figs/diff_1.png)
+![diff](./figs/diff_2.png)
+
+Note that the `trainer.train()` call is replaced by a `for` loop. At the beginning of each round (epoch in this case), we load a fixed model weights saved at the beginning, over-write the last round's record, then call `trainer.train(resume_from_checkpoint=True)` with `trainer.args.num_train_epochs` incremented by 1.
+
+So if the intended model weights (serving as the starting point for each round, the "global model" for FL use case) is properly loaded, then we shall observe a "zig-zag" pattern in the training loss curve because the model weights are reset to the same starting point at the beginning of each round. In contrast to the one-shot centralized training, where the model weights are updated continuously, and the training loss curve should follow an overall decreasing trend.
+
+To run iterative training, we use the following command:
+``` 
+python3 ./utils/hf_sft_peft_iter.py --output_path /workspace_centralized/llama2-7b-dolly-sft-iter --mode 0
+python3 ./utils/hf_sft_peft_iter.py --output_path /workspace_centralized/llama2-7b-dolly-peft-iter --mode 1
+```
+
+The PEFT curves are shown below, blue for single call, black for iterative. We can see the "zig-zag" pattern in the iterative training loss curve.
+
+![peft](./figs/cen_peft.png)
```

Comment:
Why would the graph be zig-zag?
If all states are restored properly, we should see almost match curves as training_epochs = 3?

```
trainer.train(epochs=3)
```

should be the same as

```
for i in range(3):
    trainer.train(epochs=1)
```

Reply:
I intentionally load the same model for all iterations, other wise even if the "reloading" fails, we would still see the almost match curves, and will not be able to tell.
Added some more information for this motivation